### ISSUE TITLE
Geo support

### ISSUE THREAD

**Issue by nyurik** [oxigraph/oxigraph#172]
What are your thoughts about supporting geo features?  For starters, it could be just geopoints with Wikibase-compatible distance and bounding box searches, but eventually it could also store boxes and possibly even arbitrary polygons.  I suspect this would require geometry `object` values to be stored in a separate index (not sure how Wikibase implements it), or at least to have some geo-specific index.  See also [spatial index explanation](https://postgis.net/workshops/postgis-intro/indexing.html) for PostGIS.

(P.S. awesome project, really! Thank you!!! :) )

**Comment by Tpt** [comment]
It would be amazing! It's not high on my personal priority list because I do not have a use case for it myself at the moment but if someone is willing to contribute it, I would be very please to review and merge the code.

**Comment by nyurik** [comment]
@Tpt thanks, I just began working with rust, might take me a bit of time become rust prolific))

My use case is to replace blazegraph in [sophox](https://wiki.openstreetmap.org/wiki/Sophox_query_service) service. You can see example queries at sophox.org. I think it's a perfect testing place, it has relatively low usage and low stability expectations

**Comment by Tpt** [comment]
@nyurik Amazing! I am not sure Oxigraph is ready for Sophox scale for now (I am huge fan of it, thank you!). I am currently working on a v0.3 version that improves a lot of stuff and adds a fast bulk loader. So, if you start to deep dive, the in-development `v0.3` branch is quite different from the current main branch.

**Comment by nyurik** [comment]
Thanks! I'm a bit surprised you do development in a versioned branch - usually most projects have `main` branch as the most unstable and up to date, but sure :) Bulk loader is obviously critical - blazegraph has a system for zipped ttl file processing that works fairly well, but makes my code a bit more complicated - when processing initial data I need to generate one format, whereas for updates I must use SPARQL insert queries. Not a biggie, not sure its possible to combine the two.

Why do you say Oxigraph wouldn't scale?  What kind of issues do you forsee?

**Comment by Tpt** [comment]
> Bulk loader is obviously critical - blazegraph has a system for zipped ttl file processing that works fairly well, but makes my code a bit more complicated - when processing initial data I need to generate one format, whereas for updates I must use SPARQL insert queries. Not a biggie, not sure its possible to combine the two.

Doing bulk loading when no one is reading the database is much easier: the system does not have to take care of being always consistent while loading that data. It's why there is a huge incentive of providing a separated "bulk loader".

> Why do you say Oxigraph wouldn't scale? What kind of issues do you foresee?

Sorry, I was saying that v0.2 is not scaling well at the moment, not that Oxigraph won't scale in the future.

There are two challenges with the v0.2 version:
* Loading is slow: a recent Wikidata dump will take a few days to load. The bulk loader aims to fix that.
* There is no real query optimizer and some dumb evaluation operator implementations: some queries that should be fast and are quite fast in Blazegraph are very slow. For the dumb operators, they are fixed in the v0.3 branch. For query optimization it's a huge work I leave the follow up versions (0.4...).
To summarize, the v0.1 and v0.2 version were about making Oxigraph work, v0.3 and later versions are about making Oxigraph fast.

**Comment by nyurik** [comment]
Sounds great, thanks!  Two days to ingest might be okayish, but query performance is kinda important :)  I wonder if LLVM can be used somehow to optimize the query (I'm not sure if #31 LLVM mention is relevant).  Regardless, I'm OK running the bleeding edge dev version (that's why i was mentioning `main` branch as the root) -- this way i can simply do an update&rebuild right on the prod server - worst case -- i will have to rebuild.

One other related question -- Blazegraph allows code extensions to optimize certain URI storage, e.g. the `wd:` is stored with a well known numbered prefix + the Q number, not as a long string. Does oxigraph support that?

**Comment by Tpt** [comment]
Thank you!

>  I wonder if LLVM can be used somehow to optimize the query (I'm not sure if #31 LLVM mention is relevant). 

Yes, people sometime compile CPU-heavy query parts to make them faster. PostgreSQL and Spark SQL are using it for example. But it is likly to not bring much improvements for SPARQL queries that are more I/O heavy.

> Regardless, I'm OK running the bleeding edge dev version (that's why i was mentioning main branch as the root).

Yes, it's a great idea! But don't use more "current hacking" branch like the `v0.3` branch, I do not care much about providing automated migration between individual commits in these branches.

> One other related question -- Blazegraph allows code extensions to optimize certain URI storage, e.g. the wd: is stored with a well known numbered prefix + the Q number, not as a long string. Does oxigraph support that?

Not yet but it's very easy to support. I'm going to add this soon.

**Comment by nyurik** [comment]
> Not yet but it's very easy to support. I'm going to add this soon.
Thanks, I think this is probably the biggest ask -- with Sophox, I have tons of nodes, ways, and relations, each being essentially a u64, so storing them as full strings might be mega-wasteful.

**Comment by nyurik** [comment]
Hi @Tpt , I noticed that recently a SPARQL engine newcomer - https://github.com/ad-freiburg/qlever - has been gaining a lot of attention because it hosted an example with all OpenStreetMap data (including geometries) as part of its index. I think OSM and similar projects may be a great way to popularize oxigraph index, and gain momentum.  P.S. do you know if qlever (C++) is slower/faster than oxigraph on similar tasks?

Looking forward to oneday migrate my [Sophox](https://wiki.openstreetmap.org/wiki/Sophox) project to Oxigraph :)

**Comment by Tpt** [comment]
There is now `spargeo` that implements a small part of the GeoSPARQL standard (the geof:sf* functions)

**Comment by Tpt** [comment]
Basic GeoSPARQL is there, closing into #1560
